safe_control_gym/experiments/epoch_experiment.py

```diff
@@ -0,0 +1,226 @@
+            eval_data = self._execute_task(ctrl=self.ctrl,
+                                           env=self.test_envs[test_env_num],
+                                           n_episodes=self.n_test_episodes_per_epoch,
+                                           n_steps=None,
+                                           log_freq=self.env.CTRL_FREQ)
+            if self.save_test_trajs:
+                self.add_to_all_test_data(eval_data)
+            metrics = self.compute_metrics(eval_data)
+            self.add_to_metrics(metrics)
+
+        # return learning stats
+        return_values = [deepcopy(self.metrics)]
+        if self.save_train_trajs:
+            return_values.append(deepcopy(self.all_train_data))
+        if self.save_test_trajs:
+            return_values.append(deepcopy(self.all_test_data))
+        return tuple(return_values)
+
+    def launch_single_train_epoch(self,
+                                  run_ctrl,
+                                  env,
+                                  n_episodes,
+                                  episode_i,
+                                  n_steps,
+                                  log_freq,
+                                  **kwargs):
+        """Defined per controller?"""
+        # Training Data Collection
+        traj_data = self._execute_task(ctrl=run_ctrl,
+                                       env=env,
+                                       n_episodes=n_episodes,
+                                       n_steps=n_steps,
+                                       log_freq=log_freq)
+        # Add trajectory data to all training data.
+        if self.save_train_trajs:
+            self.add_to_all_train_data(traj_data)
+        # Parsing of training Data.
+        train_inputs, train_outputs = self.preprocess_training_data(traj_data, **kwargs)
+        # Learning of training data.
+        self.train_controller(train_inputs, train_outputs, **kwargs)
+
+    def preprocess_training_data(self, traj_data, **kwargs):
+        """Preprocess the trajectory data into training data for controller training.
+
+        Args:
+            traj_data (dict): Dictionary of the trajectory data.
+
+        """
+
+        raise NotImplementedError('preprocessing_training_data not implemented!')
+
+    def train_controller(self, train_inputs, train_outputs, **kwargs):
+        """Train the controller from training data train_inputs and train_outputs. The intention is that
+        preprocess_trianing_data has processed the data and formatted such that ctrl.learn should be able to learn
+        from the given data.
+
+        Args:
+            train_inputs (np.array): Training data array.
+            train_outputs (np.array): Training data targets.
+        """
+        train_output = self.ctrl.learn(input_data=train_inputs, target_data=train_outputs)
+        raise NotImplementedError('Controller training needs to be implemented.')
+
+
+    def add_to_all_train_data(self, traj_data):
```

Comment:
Some functions missing docstrings, but not a huge deal

Reply:
Added missing docstrings.